db/view/view.cc

```diff
@@ -3576,5 +3582,453 @@ sstring build_status_to_sstring(build_status status) {
+            co_await update_building_state();
+            _as.check();
+            co_await _vb_state_machine.event.wait();
+        } catch (abort_requested_exception&) {
+        } catch (broken_condition_variable&) {
+        } catch (...) {
+            vbw_logger.warn("view_building_state_observer failed with: {}", std::current_exception());
+            sleep = true;
+        }
+
+        if (sleep && !_as.abort_requested()) {
+            try {
+                vbw_logger.debug("Sleeping after exception.");
+                co_await seastar::sleep_abortable(std::chrono::seconds(1), _as);
+            } catch (...) {
+                vbw_logger.warn("view_building_state_observer sleep failed: {}", std::current_exception());
+            }
+        }
+    }
+}
+
+future<> view_building_worker::update_building_state() {
+    auto guard = co_await _group0_client.start_operation(_as, service::raft_timeout{});
+    auto new_tasks = co_await _state.update(_db, _vb_state_machine.building_state);
+    co_await _state.finish_completed_tasks();
+    co_await start_new_tasks(std::move(new_tasks));
+}
+
+future<> view_building_worker::start_new_tasks(std::unordered_map<shard_id, std::vector<service::view_building::view_building_task>> new_tasks) {
+    for (auto& [shard, shard_tasks]: new_tasks) {
+        if (!is_shard_free(shard)) {
+            vbw_logger.warn("Tried to start new view building tasks on shard {} but the shard is busy", shard);
+            throw std::runtime_error(fmt::format("Shard {} is already busy", shard));
+        }
+
+        auto tasks = shard_tasks | std::views::transform([] (const service::view_building::view_building_task& t) {
+            return std::make_pair(t.id, t);
+        }) | std::ranges::to<std::unordered_map>();
+        auto batch = seastar::make_shared<view_building_worker::batch>(*this, tasks, shard_tasks.front().base_id, shard_tasks.front().replica);
+        for (auto& [id, _]: tasks) {
+            _state.tasks_map.insert({id, batch});
+        }
+        vbw_logger.debug("Started tasks {} for base table {} on shard {}", tasks | std::views::keys, shard_tasks.front().base_id, shard_tasks.front().replica.shard);
+        
+        co_await coroutine::maybe_yield();
+    }
+}
+
+bool view_building_worker::is_shard_free(shard_id shard) {
+    return !std::ranges::any_of(_state.tasks_map, [&shard] (auto& task_entry) {
+        return task_entry.second->replica.shard == shard;
+    });
+}
+
+void view_building_worker::init_messaging_service() {
+    ser::view_rpc_verbs::register_work_on_view_building_tasks(&_messaging, [this] (std::vector<utils::UUID> ids) -> future<std::vector<service::view_building::view_task_result>> {
+        return container().invoke_on(0, [ids = std::move(ids)] (view_building_worker& vbw) -> future<std::vector<service::view_building::view_task_result>> {
+            return vbw.get_tasks_results(std::move(ids));
+        });
+    });
```

Comment:
If the worker starts the tasks by itself and the RPC only allows to wait for the results of the tasks, perhaps we should rename the RPC accordingly. Right now it suggests as if it could start the tasks.

Reply:
changed the implementation and now the batches are started by the RPC